File: 10. Dynamic Programming/05. Best time to buy and sell stock ii/main.java
import java.util.*;

public class main {
    // 1. Brute Force (Recursive) - O(2^N)
    public static int bruteForce(int[] prices, int i, boolean holding) {
        if (i >= prices.length) return 0;

        if (holding) {
            return Math.max(prices[i] + bruteForce(prices, i + 1, false),
                            bruteForce(prices, i + 1, true));
        } else {
            return Math.max(-prices[i] + bruteForce(prices, i + 1, true),
                            bruteForce(prices, i + 1, false));
        }
    }

    public static int maxProfitBruteForce(int[] prices) {
        return bruteForce(prices, 0, false);
    }

    // 2. Memoization - O(N) Time, O(N) Space
    public static int memoizedDFS(int[] prices, int i, boolean holding, int[][] memo) {
        if (i >= prices.length) return 0;
        if (memo[i][holding ? 1 : 0] != -1) return memo[i][holding ? 1 : 0];

        if (holding) {
            return memo[i][1] = Math.max(prices[i] + memoizedDFS(prices, i + 1, false, memo),
                                         memoizedDFS(prices, i + 1, true, memo));
        } else {
            return memo[i][0] = Math.max(-prices[i] + memoizedDFS(prices, i + 1, true, memo),
                                         memoizedDFS(prices, i + 1, false, memo));
        }
    }

    public static int maxProfitMemo(int[] prices) {
        int n = prices.length;
        int[][] memo = new int[n][2];
        for (int[] row : memo) Arrays.fill(row, -1);
        return memoizedDFS(prices, 0, false, memo);
    }

    // 3. Bottom-Up DP - O(N) Time, O(N) Space
    public static int maxProfitDP(int[] prices) {
        int n = prices.length;
        int[][] dp = new int[n + 1][2];

        for (int i = n - 1; i >= 0; i--) {
            dp[i][1] = Math.max(prices[i] + dp[i + 1][0], dp[i + 1][1]);
            dp[i][0] = Math.max(-prices[i] + dp[i + 1][1], dp[i + 1][0]);
        }
        return dp[0][0];
    }

    // 4. Optimized DP - O(N) Time, O(1) Space
    public static int maxProfitOptimizedDP(int[] prices) {
        int aheadNotHolding = 0, aheadHolding = 0;

        for (int i = prices.length - 1; i >= 0; i--) {
            int tempNotHolding = Math.max(-prices[i] + aheadHolding, aheadNotHolding);
            int tempHolding = Math.max(prices[i] + aheadNotHolding, aheadHolding);
            aheadNotHolding = tempNotHolding;
            aheadHolding = tempHolding;
        }
        return aheadNotHolding;
    }

    // 5. Greedy Approach - O(N) Time, O(1) Space
    public static int maxProfitGreedy(int[] prices) {
        int profit = 0;
        for (int i = 1; i < prices.length; i++) {
            if (prices[i] > prices[i - 1]) {
                profit += prices[i] - prices[i - 1];
            }
        }
        return profit;
    }

    public static void main(String[] args) {
        int[] prices = {7, 1, 5, 3, 6, 4}; // Example input

        System.out.println("Brute Force: " + maxProfitBruteForce(prices));
        System.out.println("Memoization: " + maxProfitMemo(prices));
        System.out.println("Bottom-Up DP: " + maxProfitDP(prices));
        System.out.println("Optimized DP: " + maxProfitOptimizedDP(prices));
        System.out.println("Greedy Approach: " + maxProfitGreedy(prices));
    }
}

File: 10. Dynamic Programming/05. Best time to buy and sell stock ii/main.go
package main

import (
	"fmt"
)

// 1. Brute Force (Recursive)
func bruteForce(prices []int, i int, holding bool) int {
	if i >= len(prices) {
		return 0
	}
	if holding {
		return max(prices[i]+bruteForce(prices, i+1, false),
			bruteForce(prices, i+1, true))
	} else {
		return max(-prices[i]+bruteForce(prices, i+1, true),
			bruteForce(prices, i+1, false))
	}
}

func maxProfitBruteForce(prices []int) int {
	return bruteForce(prices, 0, false)
}

// 2. Memoization
func memoizedDFS(prices []int, i int, holding int, memo [][]int) int {
	if i >= len(prices) {
		return 0
	}
	if memo[i][holding] != -1 {
		return memo[i][holding]
	}
	if holding == 1 {
		memo[i][holding] = max(prices[i]+memoizedDFS(prices, i+1, 0, memo),
			memoizedDFS(prices, i+1, 1, memo))
	} else {
		memo[i][holding] = max(-prices[i]+memoizedDFS(prices, i+1, 1, memo),
			memoizedDFS(prices, i+1, 0, memo))
	}
	return memo[i][holding]
}

func maxProfitMemo(prices []int) int {
	n := len(prices)
	memo := make([][]int, n)
	for i := range memo {
		memo[i] = []int{-1, -1}
	}
	return memoizedDFS(prices, 0, 0, memo)
}

// 3. Bottom-Up DP
func maxProfitDP(prices []int) int {
	n := len(prices)
	dp := make([][]int, n+1)
	for i := range dp {
		dp[i] = make([]int, 2)
	}
	for i := n - 1; i >= 0; i-- {
		dp[i][1] = max(prices[i]+dp[i+1][0], dp[i+1][1]) // holding
		dp[i][0] = max(-prices[i]+dp[i+1][1], dp[i+1][0]) // not holding
	}
	return dp[0][0]
}

// 4. Optimized DP with O(1) space
func maxProfitOptimizedDP(prices []int) int {
	aheadNotHolding := 0
	aheadHolding := 0
	for i := len(prices) - 1; i >= 0; i-- {
		tempNotHolding := max(-prices[i]+aheadHolding, aheadNotHolding)
		tempHolding := max(prices[i]+aheadNotHolding, aheadHolding)
		aheadNotHolding = tempNotHolding
		aheadHolding = tempHolding
	}
	return aheadNotHolding
}

// 5. Greedy
func maxProfitGreedy(prices []int) int {
	profit := 0
	for i := 1; i < len(prices); i++ {
		if prices[i] > prices[i-1] {
			profit += prices[i] - prices[i-1]
		}
	}
	return profit
}

// Helper max
func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

// --- MAIN ---
func main() {
	prices := []int{7, 1, 5, 3, 6, 4}

	fmt.Println("Brute Force:", maxProfitBruteForce(prices))
	fmt.Println("Memoization:", maxProfitMemo(prices))
	fmt.Println("Bottom-Up DP:", maxProfitDP(prices))
	fmt.Println("Optimized DP:", maxProfitOptimizedDP(prices))
	fmt.Println("Greedy Approach:", maxProfitGreedy(prices))
}
